plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,886 @@
+        self.t_all.append(self._particle_tracker.time)
+
+
+class ParticleTracker:
+    r"""A many body particle tracker for electric and magnetic fields with no particle-particle interactions.
+
+    Particles are created and pushed through calculated E and B fields. These fields are specified as part of a grid
+    which can then be interpolated to determine local field strength. The time step used can be specified, otherwise an
+    adaptive time step is calculated based off grid resolution and the particle's gyroradius.
+
+    The simulation is terminated when a specified stop condition is met. The results of a simulation can be exported
+    using a save routine.
+
+    Parameters
+    ----------
+    grids : `~plasmapy.plasma.grids.AbstractGrid` or subclass thereof, or list
+        of same.
+        A Grid object or list of grid objects containing the required
+        quantities [E_x, E_y, E_z, B_x, B_y, B_z].
+        If any of these quantities are missing, a warning will be given and that
+        quantity will be assumed to be zero everywhere.
+
+    req_quantities : `list` of str, optional
+        A list of quantity keys required to be specified on the Grid object.
+        The default is [E_x, E_y, E_z, B_x, B_y, B_z].
+
+    verbose : bool, optional
+        If true, updates on the status of the program will be printed
+        into the standard output while running. The default is True.
+    """
+
+    def __init__(
+        self,
+        grids: Union[AbstractGrid, Iterable[AbstractGrid]],
+        req_quantities=None,
+        verbose=True,
+    ):
+        # self.grid is the grid object
+        if isinstance(grids, AbstractGrid):
+            self.grids = [
+                grids,
+            ]
+        elif isinstance(grids, collections.abc.Iterable):
+            self.grids = grids
+        elif isinstance(grids, BasePlasma):
+            raise TypeError(
+                "It appears you may be trying to access an older version of the ParticleTracker class."
+                "This class has been deprecated."
+                "Please revert to PlasmaPy version 2023.5.1 to use this version of ParticleTracker."
+            )
+        else:
+            raise TypeError("Type of argument `grids` not recognized.")
+
+        # self.grid_arr is the grid positions in si units. This is created here
+        # so that it isn't continuously called later
+        self.grids_arr = [grid.grid.to(u.m).value for grid in self.grids]
+
+        self.verbose = verbose
+
+        # This flag records whether the simulation has been run
+        self._has_run = False
+
+        self._is_adaptive_time_step = False
+        self._is_synchronized_time_step = False
+
+        # *********************************************************************
+        # Validate required fields
+        # *********************************************************************
+
+        if req_quantities is None:
```

Comment:
How should we handle required quantities?? Should it be the baseline that 

```python
req_quantities = ["E_x", "E_y", "E_z", "B_x", "B_y", "B_z"]
```

and the `req_quantities` keyword can be used to specify additional quantities?

Reply:
Good question - ultimately this depends on what other options are being used. If you are using the current pusher, you need E&B at a minimum. Once we have scattering + stopping, a density quantity will be required (along with maybe some scalar quantities like the mean ionization energy as we discussed). 

Maybe we should actually move this validation into a separate helper function, then call that in the `run` method so it use all the run settings and pick the required settings appropriately?

Reply:
Let's discuss this in our next meeting - I think it's a deeper philosophical question about what quantities go at run time (describe the run?) vs at instantiation (describe the grid?).  Either way, you cannot know just from the instantiation step what quantities are required (since that depends on the run conditions) so we would just end up moving lots of stuff to the instantiation which makes less sense to me than having it at the run method? But let's discuss.

Reply:
I've been thinking more about this - maybe the answer is to just move the specification of the termination + saving classes (and eventually whatever flags are provided to turn on stopping + scattering) to init, so we can validate the grid there?